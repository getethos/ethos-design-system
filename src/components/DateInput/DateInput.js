import React, { useState, useEffect } from 'react'
import PropTypes from 'prop-types'
import { TextMaskedInput } from '../TextMaskedInput'
import createAutoCorrectedDatePipe from 'text-mask-addons/dist/createAutoCorrectedDatePipe'
import styles from '../TextInput/TextInput.module.scss'
import errorStyles from '../Errors.module.scss'
import dayjs from '../../helpers/getDayjs.js'
import useErrorMessage from '../../hooks/useErrorMessage.js'
import useInputValidation from '../../hooks/useInputValidation.js'
import * as Validators from '../../validators/DateInputValidator'
import cleanse from '../../helpers/cleanse.js'

const { DATE_FORMATS, dateMaskByFormat, dateStringMatchesFormat } = Validators

const PrivateDateInput = (props) => {
  const {
    optional,
    dateFormat,
    allCaps,
    capitalize,
    labelCopy,
    validator,
    formChangeHandler,
    initialValue,
    currentValue,
    currentError,
    formTouched,
    setFieldTouched,
    disabled,
    guide = true,
    keepCharPositions = true,
    name = 'birthdate-auto-corrected',
    pipe = createAutoCorrectedDatePipe(dateFormat),
    mask = dateMaskByFormat[dateFormat],
    autoComplete,
    classOverrides,
    ...restProps
  } = props
  const [getError, setError, , validate] = useErrorMessage(validator)
  const val = currentValue || initialValue
  const [touched, setTouched] = useState(initialValue ? true : false)
  const [value] = useState(val || '')

  const callErrorHandlers = (value, handlerFn) => {
    const cleansed = cleanse(value)

    // Check date string format validity
    let errMsg = dateStringMatchesFormat(cleansed, dateFormat)
    const errorMessage = errMsg.length ? errMsg : ''
    if (errorMessage.length) {
      handlerFn(value, errorMessage)
    } else {
      /**
       * Note that, only once we get here, do the validators
       * get called. This is because dateStringMatchesFormat
       * is a prerequisite at the text-mask level so there's
       * no point in doing other validations if we don't even
       * have a date in valid format yet.
       */

      // Call addtional validators consumer setup (e.g. date range)
      // Note if no validators, we're still safe as validate checks
      const df = dateFormat.toUpperCase()
      const conformedDate = dayjs(value, df).format(df)
      let errorMessage = validate(conformedDate)
      errorMessage = errorMessage.length ? errorMessage : ''
      handlerFn(value, errorMessage)
    }
  }

  // Initial Value aka prefilledâ€”are considered "touched", but must prevalidate
  // which will in turn update the internal form state as to their validity
  useEffect(() => {
    if (!!formChangeHandler && initialValue) {
      doValidation(initialValue, true)
    }
  }, [])

  const [doValidation] = useInputValidation({
    validate,
    setError,
    formChangeHandler,
    callErrorHandlers,
  })

  const getClasses = () => {
    const base = `DateInput ${styles.TextInputCommon}`
    if (getError(currentError, touched)) {
      return `${base} ${errorStyles.Error}`
    }
    return classOverrides
      ? `${base} ${classOverrides}`
      : `${base} ${styles.TextInputStylable}`
  }

  return (
    <>
      <TextMaskedInput
        initialValue={value}
        optional={optional}
        mask={mask}
        pipe={pipe}
        className={getClasses()}
        type="tel"
        labelCopy={labelCopy}
        labelColor={props.labelColor}
        labelWeight={props.labelWeight}
        labelClasses={props.labelClasses}
        allCaps={allCaps}
        capitalize={capitalize}
        data-tid={restProps['data-tid']}
        guide={guide}
        doValidation={doValidation}
        name={name}
        placeholder={dateFormat}
        keepCharPositions={keepCharPositions}
        currentValue={currentValue}
        currentError={currentError}
        formTouched={formTouched}
        setFieldTouched={setFieldTouched}
        getTouched={touched}
        setTouched={setTouched}
        disabled={disabled}
        autoComplete={autoComplete}
        classOverrides={classOverrides}
      />
      {getError(currentError, touched)}
    </>
  )
}

PrivateDateInput.PUBLIC_PROPS = {
  optional: PropTypes.bool,
  dateFormat: PropTypes.oneOf(DATE_FORMATS),
  'data-tid': PropTypes.string.isRequired,
  disabled: PropTypes.bool,
  allCaps: PropTypes.bool,
  /** text transform capitalize label */
  capitalize: PropTypes.bool,
  name: PropTypes.string.isRequired,
  labelCopy: PropTypes.string.isRequired,
  validator: PropTypes.func,
  initialValue: PropTypes.string,
  guide: PropTypes.bool,
  keepCharPositions: PropTypes.bool,
  autoComplete: PropTypes.string,
  classOverrides: PropTypes.string,
  labelColor: PropTypes.string,
  labelWeight: PropTypes.string,
  labelClasses: PropTypes.string,
}

PrivateDateInput.propTypes = {
  ...PrivateDateInput.PUBLIC_PROPS,
}

PrivateDateInput.defaultProps = {
  dateFormat: 'mm/dd/yyyy',
  labelCopy: 'Date',
}

const DateInputFactory = (privateProps) => {
  const PublicDateInputComponent = (downstreamProps) => {
    return <PrivateDateInput {...downstreamProps} {...privateProps} />
  }

  return PublicDateInputComponent
}

export const DateInput = DateInputFactory()

export const DateInputValidators = Validators
