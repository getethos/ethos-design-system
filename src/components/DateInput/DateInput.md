```jsx
var utc = require('dayjs/plugin/utc')

import dayjs from '../../helpers/getDayjs.js'
import validateExists from '../../validators/validateExists'
import {
  getMinDateValidator,
  getMaxDateValidator,
} from '../../validators/DateInputValidator'

import { TitleLarge2, Form, Spacer, Button, InfoMessage } from '../index'

const MIN_ERROR_MESSAGE = "That's too far in the past!"
const minDate = dayjs('1/1/1700', 'M/D/YYYY')
  .utc()
  .startOf('day')
  .toDate()
const minDateValidator = getMinDateValidator({
  minDate,
  customErrorMessage: MIN_ERROR_MESSAGE,
  dateFormat: 'mm/yyyy',
})

const MAX_ERROR_MESSAGE = "That's too far in the future!"
const maxDate = dayjs()
  .utc()
  .startOf('day')
  .add('10', 'years')
  .toDate()
const maxDateValidator = getMaxDateValidator({
  maxDate,
  customErrorMessage: MAX_ERROR_MESSAGE,
  dateFormat: 'mm/yyyy',
})

;<Form
  config={{
    formName: 'Prefilled w/DateInput',
    formId: '1',
    fields: {
      last_seen_date: {
        component: (props, options) => {
          return <DateInput dateFormat="mm/yyyy" {...props} />
        },
        labelCopy: ' ',
        tid: 'dateinput-prefilled-tid',
        validators: [validateExists, minDateValidator, maxDateValidator],
      },
    },
    onSubmit: async (formData) => {},
  }}
>
  {(api) => {
    const {
      field,
      getFormErrorMessage,
      getFormIsValid,
      getFormInteractedWith,
    } = api
    return (
      <div>
        <TitleLarge2.Sans.Regular400 element="label">
          When were you last seen by a Doctor?
        </TitleLarge2.Sans.Regular400>
        <Spacer.H16 />
        {field('last_seen_date')}
        <Spacer.H16 />
        <Button.Medium.Black disabled={!getFormIsValid()} type="submit">
          Continue
        </Button.Medium.Black>
      </div>
    )
  }}
</Form>
```
