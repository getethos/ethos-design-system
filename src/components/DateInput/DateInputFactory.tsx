import * as React from 'react'
import { PrivateDateInput } from './PrivateDateInput'
export const DateInputFactory = (privateProps: any) => {
  const PublicDateInputComponent = (downstreamProps: any) => {
    return <PrivateDateInput {...downstreamProps} {...privateProps} />
  }

  return PublicDateInputComponent
}
