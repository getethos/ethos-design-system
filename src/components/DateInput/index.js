export { DateInput } from './DateInput.js'
