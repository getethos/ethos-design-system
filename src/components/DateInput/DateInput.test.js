import React from 'react'
import { DateInput } from './DateInput'
import renderer from 'react-test-renderer'

// TODO: make a helper
// https://github.com/text-mask/text-mask/issues/427
jest.mock('react-text-mask', () => (props) => <input type="text" {...props} />)

describe('DateInput', () => {
  it('default rendering', () => {
    const tree = renderer
      .create(
        <DateInput
          dateFormat="mm/yyyy"
          labelCopy="When did you apply for this life insurance?"
          data-tid="when-applied-for"
          validator={() => {}}
          name="whenapplied"
        />
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
