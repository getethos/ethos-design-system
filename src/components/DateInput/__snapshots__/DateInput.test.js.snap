// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateInput default rendering 1`] = `
Array [
  <label
    className="Caption Theinhardt Medium500 GrayPrimary AllCaps"
    htmlFor="whenapplied"
  >
    When did you apply for this life insurance?
  </label>,
  <div
    aria-hidden={true}
    className="Spacer"
    style={
      Object {
        "height": 4,
      }
    }
  />,
  <input
    autoComplete="date"
    className="TextMaskedInput TextInput"
    data-tid="when-applied-for"
    disabled={false}
    guide={true}
    keepCharPositions={true}
    mask={
      Array [
        /\\\\d/,
        /\\\\d/,
        "/",
        /\\\\d/,
        /\\\\d/,
        /\\\\d/,
        /\\\\d/,
      ]
    }
    name="whenapplied"
    onBlur={[Function]}
    onChange={[Function]}
    pipe={[Function]}
    placeholder="mm/yyyy"
    type="tel"
    value=""
  />,
  "",
]
`;
